From 29171b4545830323999c463a30a2e0ffed297c2f Mon Sep 17 00:00:00 2001
From: Philip Taron <philip@qumulo.com>
Date: Thu, 19 Mar 2026 16:58:02 -0700
Subject: [PATCH 4/7] Port patch 16: Add krb5_set_debug_dest_facility()

Add public API function to set the debug logging facility for a
krb5_context, allowing custom logger callbacks into the application's
logging system.
---
 lib/krb5/libkrb5-exports.def.in |  1 +
 lib/krb5/log.c                  | 15 +++++++++++++++
 lib/krb5/version-script.map     |  1 +
 3 files changed, 17 insertions(+)

diff --git a/lib/krb5/libkrb5-exports.def.in b/lib/krb5/libkrb5-exports.def.in
index 191a0c48c..6e14eb5d6 100644
--- a/lib/krb5/libkrb5-exports.def.in
+++ b/lib/krb5/libkrb5-exports.def.in
@@ -617,6 +617,7 @@ EXPORTS
 	krb5_set_config
 	krb5_set_config_files
 	krb5_set_debug_dest
+	krb5_set_debug_dest_facility
 	krb5_set_default_in_tkt_etypes
 	krb5_set_default_realm
 	krb5_set_dns_canonicalize_hostname
diff --git a/lib/krb5/log.c b/lib/krb5/log.c
index 306431a5c..125428a55 100644
--- a/lib/krb5/log.c
+++ b/lib/krb5/log.c
@@ -132,6 +132,21 @@ krb5_closelog(krb5_context context,
     return 0;
 }
 
+/**
+ * Set the debug logging facility.
+ *
+ * @param context A Kerberos 5 context
+ * @param fac Facility to use for logging.
+ *
+ * @ingroup krb5_error
+ */
+
+KRB5_LIB_FUNCTION krb5_error_code KRB5_LIB_CALL
+krb5_set_debug_dest_facility(krb5_context context, krb5_log_facility *fac)
+{
+    return heim_set_debug_dest(context->hcontext, fac);
+}
+
 #undef __attribute__
 #define __attribute__(X)
 
diff --git a/lib/krb5/version-script.map b/lib/krb5/version-script.map
index 2b39d63a1..ac01d5a5d 100644
--- a/lib/krb5/version-script.map
+++ b/lib/krb5/version-script.map
@@ -610,6 +610,7 @@ HEIMDAL_KRB5_2.0 {
 		krb5_set_config;
 		krb5_set_config_files;
 		krb5_set_debug_dest;
+		krb5_set_debug_dest_facility;
 		krb5_set_default_in_tkt_etypes;
 		krb5_set_default_realm;
 		krb5_set_dns_canonicalize_hostname;
-- 
2.53.0

